File: bent/magic_square.lua
-- Attempt to solve the prime number magic
-- square problem in Lua
--
-- Magic square (4x4) composed of prime numbers in the range 1-99.
-- All row totals = 202
-- All col totals = 202
-- Both diag totals = 202
--


valid_grid = function(b)
    local x
    local c
    local prevj
    local d1
    local d2
    -- check to make sure there are 16 unique values
    x={}
    for i1,j1 in ipairs(b) do
        for i2,j2 in ipairs(j1) do
            table.insert(x,j2)
        end
    end
    table.sort(x)
    prevj=-1
    for i,j in ipairs(x) do
        if (j==prevj) then
            return(false)
        end
        prevj=j
    end
    -- check to make sure each column sum is 202
    for c=1,4 do
        if ((b[1][c]+b[2][c]+b[3][c]+b[4][c]) ~= 202) then
            return(false)
        end
    end
    -- check to make sure each diag sum is 202
    d1=b[1][1]+b[2][2]+b[3][3]+b[4][4]
    if (d1 ~= 202) then
        return(false)
    end
    d2=b[4][1]+b[3][2]+b[2][3]+b[1][4]
    if (d2 ~= 202) then
        return(false)
    end
    return(true)
end

dump = function(b)
    for i,j in ipairs(b) do
        for i2,j2 in ipairs(j) do
            io.write("\t" .. j2)
        end
        print()
    end
end

match_start = function(a,x)
    for i,j in ipairs(x) do
        if (a[i] ~= j) then
            return(false)
        end
    end
    return(true)
end


p = {2, 3, 5, 7, 11,13,17,19,23,29,31,37,
    41,43,47,53, 59,61,67,71,73,79,83,89,
    97}

-- master list of feasible rows
a={}
for i1,j1 in ipairs(p) do
    for i2,j2 in ipairs(p) do
        for i3,j3 in ipairs(p) do
            for i4,j4 in ipairs(p) do
                if ((j1 ~= j2) and (j1 ~= j3) and (j1 ~= j4)) then
                    if ((j2 ~= j3) and (j2 ~= j4) and (j3 ~= j4)) then
                        if ((j1+j2+j3+j4)==202) then
                            table.insert(a,{j1,j2,j3,j4})
                        end
                    end
                end
            end
        end
    end
end

-- this should be a list of 3120 permutations
print("Number of feasible rows = ",#a)

-- try each as the first row of the grid
b={{0,0,0,0},{0,0,0,0},{0,0,0,0},{0,0,0,0}}

for i1,j1 in ipairs(a) do -- i1 points to first row
    b[1]=a[i1]
    print("Testing: ",b[1][1],b[1][2],b[1][3],b[1][4])
    for i2,j2 in ipairs(a) do -- i2 points to first col
        if ((i2 ~= i1) and match_start(j2,{b[1][1]})) then
            b[2][1]=j2[2]; b[3][1]=j2[3]; b[4][1]=j2[4]
            for i3,j3 in ipairs(a) do -- i3 points to second row
                feasible=match_start(j3,{b[2][1]})
                if ((i3 ~= i1) and (i3 ~= i2) and feasible ) then
                    b[2]=a[i3]
                    for i4,j4 in ipairs(a) do -- i4 points to second column
                        feasible=match_start(j4,{b[1][2],b[2][2]})
                        if ((i4~=i3) and (i4~=i2) and (i4~=i1) and feasible) then
                            b[3][2]=j4[3]; b[4][2]=j4[4]
                            for i5,j5 in ipairs(a) do -- i5 points to third row
                                feasible=match_start(j5,{b[3][1],b[3][2]})
                                if ((i5~=i4) and (i5~=i3) and (i5~=i2) and (i5~=i1) and feasible) then
                                    b[3]=a[i5]
                                    for i6,j6 in ipairs(a) do -- i6 points to third col
                                        feasible=match_start(j6,{b[1][3],b[2][3],b[3][3]})
                                        feasible=feasible and (i6~=i5) and (i6~=i4) and (i6~=i3)
                                        feasible=feasible and (i6~=i2) and (i6~=i1)
                                        if (feasible) then
                                            b[4][3]=j6[4]
                                            for i7,j7 in ipairs(a) do -- i7 points to 4th row
                                                feasible=match_start(j7,{b[4][1],b[4][2],b[4][3]})
                                                feasible=feasible and (i7~=i6) and (i7~=i5) and (i7~=i4)
                                                feasible=feasible and (i7~=i3) and (i7~=i2) and (i7~=i1)
                                                if (feasible) then
                                                    if (valid_grid(b)) then
                                                        print("Solution:")
                                                        dump(b)
                                                    end
                                                end
                                            end
                                        end
                                    end
                                end
                            end
                        end
                    end
                end
            end
        end
    end
end

File: bent/magic_square.R
library(gtools)

p=read.table('primes.txt')
p=p[,1]
p=p[p<100]

a=permutations(25,4,p)
s=apply(a,1,sum)
a=a[s==202,]

print(paste0("size of matrix = ",dim(a)))

# Monte Carlo attempt to get lucky 
# No joy after 31e6 tries
if (F) {
    t1=Sys.time()
    for (i in 1:10000000) {
        s=sample(nrow(a),4)
        b=a[s,]
        if (all(apply(b,2,sum)==202)) {
            if (length(unique(as.vector(b)))==16) {
                print(b)
                break
            }
        }
    }
    t2=Sys.time()
    print(t2-t1)
}

# For each possible row in a
# Pick a potential set for the first column
# For each first row/column pick a potential set for second row
# Pick potential set for second column
# etc. until you find a solution or find zero solutions



# Get a row suitable for inclusion in the b matrix
# at a position given by rownum.  The first colnum
# columns must match the values in b[rownum,].
get_row = function(a,b,rownum,colnum) {
    # match the leading column values
    ### s=(a[,1:(colnum-1)]==b[rownum,(1:colnum-1)])
    s=apply(a,1,function(x) {all(x[1:(colnum-1)]==b[rownum,(1:colnum-1)])})
    c=matrix(a[s,],ncol=4)
    # eliminate rows that duplicate existing values
    s=apply(c,1,function(x) {!any(x[colnum:4] %in% b)})
    c=matrix(c[s,],ncol=4)
    return(c)
}

# Get a column suitable for inclusion in the b matrix
# at a position given by colnum.  The first rownum
# values must match the values in b[,colnum].
get_col = function(a,b,rownum,colnum) {
    # match the leading row values
    ### s=(a[,1:(rownum-1)]==b[1:(rownum-1),colnum])
    s=apply(a,1,function(x) {all(x[1:(rownum-1)]==b[1:(rownum-1),colnum])})
    c=matrix(a[s,],ncol=4)
    # eliminate rows that duplicate existing values
    s=apply(c,1,function(x) {!any(x[rownum:4] %in% b)})
    c=matrix(c[s,],ncol=4)
    return(c)
}

# At any point the b matrix contains a partial solution
#
b=matrix(rep(0,16),ncol=4)

# Starting at row 247 based on the known solution
# Row 247 should be a solution

for (i1 in 247:nrow(a)) {
    print(paste('Status = ',paste(a[i1,],collapse=' ')))
    b[1,]=a[i1,]
    c=get_col(a,b,2,1)
    if (nrow(c)==0) {
        next
    }
    b2=b
    for (i2 in 1:nrow(c)) {
        b2[,1]=c[i2,]
        d=get_row(a,b2,2,2)
        if (nrow(d)==0) {
            next
        }
        b3=b2
        for (i3 in 1:nrow(d)) {
            b3[2,]=d[i3,]
            # Check the first two elements on the diagonal
            s=apply(a,1,function(x) {all(x[1:2]==c(b3[1,1],b3[2,2]))})
            if (sum(s)==0) {
                next
            }
            # Check the last two elements on the anti-diagonal
            s=apply(a,1,function(x) {all(x[3:4]==c(b3[2,3],b3[1,4]))})
            if (sum(s)==0) {
                next
            }
            e=get_col(a,b3,3,2)
            if (nrow(e)==0) {
                next
            }
            b4=b3
            for (i4 in 1:nrow(e)) {
                b4[,2]=e[i4,]
                f=get_row(a,b4,3,3)
                if (nrow(f)==0) {
                    next
                }
                b5=b4
                for (i5 in 1:nrow(f)) {
                    b5[3,]=f[i5,]
                    # Check the first three elements on the diagonal
                    s=apply(a,1,function(x) {all(x[1:3]==c(b5[1,1],b5[2,2],b5[3,3]))})
                    if (sum(s)==0) {
                        next
                    }
                    # Check the last three elements on the anti-diagonal
                    # This line originally had a typo that prevented finding
                    # a solution.
                    s=apply(a,1,function(x) {all(x[2:4]==c(b5[3,2],b5[2,3],b5[1,4]))})
                    if (sum(s)==0) {
                        next
                    }
                    g=get_col(a,b5,4,3)
                    if (nrow(g)==0) {
                        next
                    }
                    b6=b5
                    for (i6 in 1:nrow(g)) {
                        b6[,3]=g[i6,]
                        h=get_row(a,b6,4,4)
                        if (nrow(h)==0) {
                            next
                        }
                        b7=b6
                        for (i7 in 1:nrow(h)) {
                            b7[4,]=h[i7,]
                            # Feasible Solution -- now check the diagonals
                            if (sum(diag(b7)) != 202) {
                                next
                            }
                            s=b7[1,4]+b7[2,3]+b7[3,2]+b7[4,1]
                            if (s != 202) {
                                next
                            }
                            print("Solution:")
                            print(b7)
                        }
                    }
                }
            }
        }
    }
}

